### ISSUE TITLE
[FEATURE] Run Cassandra migration tool on startup

### ISSUE THREAD

**Issue by Burdu** [jhipster/generator-jhipster#4925]
I put some code together to execute the Cassandra migration tool on startup for dev purposes. It works on dev profile currently, but can be tweaked to work for prod as well.
It is a bit low level, but does the job.

##### **Related issues**

https://github.com/jhipster/generator-jhipster/pull/3593

##### **Suggest a Fix**

    @Configuration
    @EnableConfigurationProperties(CassandraProperties.class)
    @Profile({Constants.SPRING_PROFILE_DEVELOPMENT, Constants.SPRING_PROFILE_PRODUCTION})
    public class CassandraConfiguration {

        @Value("${spring.data.cassandra.protocolVersion:V4}")
        private ProtocolVersion protocolVersion;

        @Autowired(required = false)
        MetricRegistry metricRegistry;

        private final Logger log = LoggerFactory.getLogger(CassandraConfiguration.class);

        @Bean
        public Cluster cluster(CassandraProperties properties) throws IOException, InterruptedException {
            runMigration();
            // other code in the cluster bean instantiation
        }

        // other configuration code for cassandra

        private void runMigration() throws IOException, InterruptedException {
            log.debug("Running Cassandra migration tool");
            ProcessBuilder pb = new ProcessBuilder(
                "src/main/docker/cassandra/scripts/autoMigrate.sh",
                "src/main/resources/config/cql/changelog/");
            Map<String, String> systemEnvironment = pb.environment();
            systemEnvironment.put("CASSANDRA_CONTACT_POINT", "127.0.0.1");
            Process process = pb.start();
            try (BufferedReader processOutput = new BufferedReader(new InputStreamReader(process.getInputStream()))) {
                String line;
                while ((line = processOutput.readLine()) != null) {
                    log.debug(line);
                }
            }
            int exitValue = process.waitFor();
            if (exitValue != 0) {
                try (BufferedReader processOutput = new BufferedReader(new InputStreamReader(process.getErrorStream()))) {
                    String line;
                    while ((line = processOutput.readLine()) != null) {
                        log.debug(line);
                    }
                }
                throw new RuntimeException("Failed to execute cassandra migration script!");
            }
        }
    }

**Comment by jdubois** [comment]
Since the first time I worked on this, it was totally refactored by @raphaelbrugier -> Raphaël, what do you think about this?

**Comment by raphaelbrugier** [comment]
Yes, this could make sense in development, having the scripts automatically executed when you restart the application.

I would avoid doing this in production.
The main problem is that if you are deploying multiple instances of a new version, there is nothing preventing the script to be run in parallel on each instance. Unlike liquibase which has a lock mechanism.

Notes that the script has some [requirements](https://jhipster.github.io/using-cassandra/#prerequisites), you may want to check these before running the script to give the user a proper error.

I can help you reviewing the PR.

**Comment by jdubois** [comment]
Oh yes thanks @raphaelbrugier I forgot about the locking issue!!

I'm sorry @Burdu but we just can't do this: without a good locking mechanism, that's going to fail in many cases. Of course you could only have it in dev mode, but then it's a bit weird to have it work differently in dev and prod.

Another issue I have is that you run a `.sh` script in your app, so it probably will fail on Windows.

It's not that complex to run the upgrade script manually outside of the application, so for me this isn't a big issue anyway.

**Comment by Burdu** [comment]
I understand.
Please note that you still have no locking mechanism if you start multiple docker instances sharing the same volume for cassandra data
